Atdd Step4 리뷰 요청드립니다.

deokmoon opened the pull request:
안녕하세요 영훈님 :)

인수테스트주도개발 과정 마지막 미션 1차 진행하여 리뷰 요청드립니다.
영훈님의 꼼꼼한 리뷰 덕분에 많이 배웠다고 생각하는데, 벌써 마지막 미션이라 생각하니 많이 아쉽네요.
다시 한번 감사드립니다. 👍 

이번 미션은 ATDD의 힘을 느끼며 미션을 진행했습니다. 

처음 ATDD를 접했을 때 ATDD가 시나리오 중심으로 작성해야된다고 배웠는데,
시나리오를 테스트코드로 어떻게 풀어가야할지 고민이 됐었습니다.

Step4 미션은, 미션에 대한 기능사항 및 예외사항을 기반으로 요구사항을 작성하고, 
RestAssured 기반으로 테스트코드로 요구사항을 테스트하니 작성한 코드에 대한 신뢰를 할 수 있었습니다. 

JPA를 실무에서 사용해보지 않고, 혼자 조금 학습한 정도여서 이렇게 기능을 구현해도 되는건가 ?  라는 의심을 갖고 미션을 진행했었는데, 
시나리오대로 인수테스트 하나씩 작성하며 기능을 구현하고 log를 통해 쿼리를 확인하니 작성한 Product 코드에 대해 조금은 신뢰를 갖을 수 있었습니다. (아 ~ 이래서 ATDD를 하는구나 하면서 미션을 진행했습니다 ㅎㅎ..)

사실 살짝 미션에 대한 권태가 왔었는데, 다시 한번 흥미를 일꺠우는 미션이 되었습니다!
다시 한번 상세한 리뷰 해주신 영훈님께 감사드리며, 이번 미션도 리뷰 잘 부탁드리겠습니다.

감사합니다.😊

Commit:
refactor: step3 리뷰 피드백 적용 리팩토링

deokmoon committed:
refactor: 리뷰 피드백 적용
 - ifPresent 적용
 - jpa 중복 내용 제거(createStation)

deokmoon committed:
docs: step3 회고 및 step4 요구사항 작성

deokmoon committed:
test: 삭제 요청 파라미터에 대한 검증 인수테스트 작성

deokmoon committed:
feat: 구간삭제 요청 API controller, service 구현

deokmoon committed:
feat: 삭제 요청 파라미터에 대한 검증로직 적용

deokmoon committed:
Test: 종점 구간 제거 인수 테스트 코드 작성

deokmoon committed:
feat: 종점 구간 제거 기능 구현

deokmoon committed:
docs: 기능 구현사항 현행화

deokmoon committed:
test: 중간 구간 제거 인수테스트 작성

deokmoon committed:
feat: 중간 구간 제거 기능 구현

deokmoon committed:
docs: 기능 요구사항 현행화 - 가운데 구각 제거 구현 완료

chae-yh commented on README.md:
```diff
@@ -220,3 +220,59 @@ host: localhost:52165
   * 참고 코드: https://github.com/next-step/atdd-subway-map/blob/boorownie/src/main/java/nextstep/subway/line/domain/LineStations.java
   * JPA @Embedded And @Embeddable 문서 참고: https://www.baeldung.com/jpa-embedded-embeddable
   
+#### Step3 회고
```
회고 작성 좋네요 👍 저도 보고 배우고 있습니다 ㅎㅎ

chae-yh commented on README.md:
```diff
@@ -220,3 +220,59 @@ host: localhost:52165
+  * 지연로딩 설정한 연관관계에서 proxy 객체로 존재하는 경우가 있음
+  * 프록시는 해당 Entity 기반의 인스턴스는 맞으나, 해당 Entity의 클래스가 아님(오역 발생 여부 있음)
+  * 따라서 ob.getClass() 비교 및 obj instanceof Entity클래스 를 모두 해야 함
+  * hibernate는 equals()와 hashcode() 구현을 추천하지 않는다고 하는데, 이 내용만 보고 판단하는 것이 아니라 좀 더 찾아볼 필요성이 있음(JPA 스터디 진행)
```
오.. 이 내용은 저도 정확히는 몰랐는데, 배우게 되었습니다 👍 
한가지 궁금한 것은 equals 사용 시 id비교가 아니라 두 entity간 unique한 특정값으로 비교하는 것이 좋다.. 고 이해했는데, 혹시 작성해주신 부분은 다른 부분일까요...? 
 -> `hibernate는 equals()와 hashcode() 구현을 추천하지 않는다고 하는데`
 -> `Hibernate recommends not implementing equals() and hashCode() by using the ID, but rather by using a natural identifier. Implementing it with IDs can cause problems because entities don't have an ID until they're saved and the ID is generated`

deokmoon replied:
네넵 말씀해주신 내용이 맞습니다. 

이 스택오버플오우 내용  떄문에 많이 헷갈렸습니다. 
(저기 답변 해주신 분이 표준 문서에서 발췌한 것처럼 잘 포장해서 넣어서..)

마치 스택오버플로우 답변 내용이, hibernate 표준 문서에 ID 값이 유니크한 값일 때도 equals()와 hashcode()  구현 시 id로만 비교하는 것을 추천하지 않는다 라고 받아들여 오해가 생긴 것 같습니다. 

오해를 하니 ID가 유니크한데 굳이 다른 값으로 엔티티 식별자로 설정해서 equals()와 hashcode()를 구현해야하나 ? 라는 생각이 들었습니다.

덕분에 표준 문서, 영한님의 JPA책, 우테캠 채널 등을 찾아보았습니다. ID 값이 유니크하다면 충분히 equals()와 hashcode() 를 구현해서 사용해도 되는 것으로 확인했습니다.

-https://docs.jboss.org/hibernate/stable/core.old/reference/en/html/persistent-classes-equalshashcode.html

chae-yh commented on README.md:
```diff
@@ -220,3 +220,59 @@ host: localhost:52165
+  ~~~
+  
+  
+### Step4 구간 제거 기능
```
기능 목록 작성 👍

chae-yh commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -16,15 +15,17 @@
 public class LineService {
     private LineRepository lineRepository;
     private StationRepository stationRepository;
+    private SectionRepository sectionRepository;
```
오.. SectionRepository를 추가해주셨군요...! 혹시 Sections에 CascadeType.ALL로 주신 것이 동작을 안해서 넣으신 걸까요...?

deokmoon replied:
우선은 Sections의 orphanRemoval = true 를 설정 후 SectionRepository 는 제거하였습니다.

사실 삭제 로직을 구현하면서 고민했던 부분입니다.

remove 에 대한 비즈니스 로직 처리할 때 고아객체 설정하여,
remove -> 연관되어 있는 엔티티 같이 delete 하도록 설정하는 것과

최종적으로 id를 가져와, 
최종적으로 service 레이어에서 삭제 대상 id로 repository 의 메서드를 호출하는 것 중
어느 방법이 더 팀원들에게 명확할까를 고민했었습니다. 
(요청은 station으로 요청하지만, 삭제는 section이기 때문에)

고민끝에 삭제 전 Id로 날리는 것이 더 명확해 보여서 SectionRepository  를 만들어, id를 리턴 받아 처리하였습니다. 여기서 명확하다는 표현은 orphanRemoval 설정을 통해 JPA 내부적으로 작동하는 것보다, service layer에서 id로 삭제 하는 것이 코드로 보이기 때문에 그렇게 표현했습니다. (지극히 주관적인 내용이었습니다 ㅎㅎ)

JPA를 실무에서 많이 사용하지 못하여 위와 같이 생각했는데,
(이제 레거시 개선 프로젝트하면서 적용 중입니다.)
리뷰 주신 내용을 생각해보니, 현업에서는 id 리턴 보다는 최대한 JPA 동작을 활용하는 것 같아 수정하였습니다!

chae-yh commented on src/main/java/nextstep/subway/domain/Line.java:
```diff
@@ -55,6 +55,16 @@ public void addSection(Section section) {
         this.sections.add(section);
     }
 
+    public long deleteStation(Station reqDeleteStation) {
```
req와 같은 축약어보다는 전체 이름을 작성해보는 것도 좋을 것 같아요..!

chae-yh commented on src/main/java/nextstep/subway/domain/Line.java:
```diff
@@ -55,6 +55,16 @@ public void addSection(Section section) {
+        }
+        if (sections.hasNotStation(reqDeleteStation)) {
+            throw new IllegalArgumentException(ErrorCode.NO_SUCH_STATION_IN_THE_LINE_EXCEPTION.getErrorMessage());
+        }
```
여기의 두가지 validation은 Sections에서 해줘도 되지 않을까요...? Line보다는 Sections의 책임이 적합해 보여 여쭤봅니다...!

deokmoon replied:
오 넵
말씀해주신대로 Sections에 메세지 보내서 sections에서 처리하는 방식이 더 객체지향스러워보입니다!

chae-yh commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -178,4 +179,48 @@ private boolean isSectionsNotEmpty() {
         return this.sections.size() == ZERO;
     }
+
+    public boolean isDeleteUnable() {
```
이건 전혀 중요한 것은 아닙니다만...!
isUnableToDelete가 문법적으로 맞지 않을까요...! ㅎ.ㅎ;;;;;;;

deokmoon replied:
엇 ㅋㅋㅋㅋㅋ... 맞습니다 ... 당황스럽네요..
isUnableToDelete 로 생각한거 같은데 다른 메서드명으로 커밋되어 있네요 ㅎㅎ

chae-yh commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -178,4 +179,48 @@ private boolean isSectionsNotEmpty() {
+        Section lastSection = findLastSection();
+        if (firstSection.isSameUpStation(reqDeleteStation)) {
+            this.sections.remove(firstSection);
+            return firstSection.getId();
```
이 부분은 재밌네요 ㅎ.ㅎ! 혹시 해당 id를 리턴해주시는 이유를 여쭤봐도 될까요...? 실제로 api response로는 사용되고 있지 않은 것 같아 여쭤봅니다..!

deokmoon replied:
sectionRepository 로 remove  처리 했던 사유와 같습니다!

chae-yh commented on src/test/java/nextstep/subway/Section/SectionsAcceptanceTest.java:
```diff
@@ -117,6 +119,87 @@ Stream<DynamicTest> makeExceptionWhenMatchBothStation() {
         );
     }
 
+    @DisplayName("구간삭제 요청 파라미터에 대한 예외 검증 - LINE이 존재하지 않은 경우")
```
인수테스트 작성 👍

chae-yh commented on src/test/java/nextstep/subway/Section/SectionsAcceptanceTest.java:
```diff
@@ -117,6 +119,87 @@ Stream<DynamicTest> makeExceptionWhenMatchBothStation() {
+        assertThat(response.statusCode()).isEqualTo(HttpStatus.INTERNAL_SERVER_ERROR.value());
+    }
+
+    @DisplayName("구간삭제 요청 파라미터에 대한 예외 검증 - 삭제 요청 Station이 요청한 Line에 등록되지 않은 경우")
```
다이나믹 테스트도 잘 만들어 주셨네요 👍

chae-yh commented on src/test/java/nextstep/subway/Section/SectionsTest.java:
```diff
@@ -17,7 +16,6 @@
 import static org.junit.jupiter.api.Assertions.assertAll;
 
-@DataJpaTest
 class SectionsTest {
```
Sections에 많은 로직이 들어갔는데, 단위 테스트를 추가해서 작성해보면 어떨까요...?

chae-yh reviewed (changes requested):
덕문님 안녕하세요. 벌써 마지막 step이네요 ㅠ 그리고 전체적으로 코드도 잘 작성해주셨습니다 👍 다만 조금만 더 고민해주시면 좋을만한 내용들이 있어 코멘트 드리니 확인 부탁드립니다.

감사합니다 : )

deokmoon committed:
feat: remove line station 처리 시 응답값에 Line이 가진 Stations 추가

deokmoon committed:
test: line station remove 인수테스트 조회 시나리오 추가 적용

deokmoon committed:
refactor: line station remove 요청 시 orphanRemoval = true 적용

 - 기존 id를 리턴 받아 repository 인터페이스를 구현하여 삭제
 - orphanRemoval 를 통해 영속성 적용 및 고아객체 적용하여 sections.remove() 시 삭제되도록 적용

deokmoon committed:
test: sections 일급콜렉션 단위테스트 케이스 추가 적용
- 구간 삭제 확인
- 예외사항 테스트코드 작성 - 구간이 하나일 때, 노선에없는 station 요청할 때

deokmoon committed:
docs: equals 추가 확인한 내용 작성하여 문서 현행화

chae-yh reviewed (approved):
덕문님 안녕하세요. 피드백 잘 반영해주셨네요 👍 더 이상 특별히 코멘트 드릴 내용이 없네요 ㅎㅎㅎ merge하겠습니다...!

pr 중에 코멘트 주신 내용을 보니 해당 미션이 덕문님께 긍정적인 효과를 준 것 같아 다행이라는 생각이 듭니다. 미션 중에도 피드백을 많이 드린 것 같은데, 잘 반영해주셔서 감사했습니다. 나름대로는 열심히 피드백 드리려고 노력했는데, 덕문님의 성장에 조금이라도 도움이 되었다면 좋겠습니다 ㅎㅎ
이제 코스도 마지막을 향해 달려가고 있는데, 남은 미션에서도 이렇게 즐겁게 성장해주시길 빌겠습니다. 그리고 꼭 코스도 완주하시길 빕니다...!

그 동안 고생 많으셨습니다. 감사합니다 : )
